phpseclib/Net/SSH2.php

```diff
@@ -3457,6 +3458,10 @@ private function get_binary_packet($skip_channel_filter = false)
                     $remaining_length = 0;
                     break;
                 case 'chacha20-poly1305@openssh.com':
+                    if (!($this->decrypt instanceof Salsa20)) {
+                        throw new \InvalidArgumentException('$this->decrypt is not a ' . Salsa20::class);
+                    }
+
```

Comment:
We do not have to check `$this->decrypt instanceof Rijndael` because we are not using any `Rijndael` specific methods.

But in the `chacha20-poly1305@openssh.com`, we are using a `ChaCha20` specific method: `setCounter()`. This is why we have to narrow the type using an instanceof check.

Here is the error:
```txt
ERROR: UndefinedMethod - ../phpseclib/Net/SSH2.php:3478:37 - Method phpseclib3\Crypt\Common\SymmetricKey::setCounter does not exist (see https://psalm.dev/022)
                    $this->decrypt->setCounter(0);
```

To address your redundancy concern, we can refactor the code to not use `$this->decryptName` by doing something like `$this->decrypt instanceof Rijndael && $this->decrypt->getMode() === 'mode'`.

Reply:
It's still not clear to me how `$this->decryptName` could be chacha20-poly1305@openssh.com while `$this->decrypt` _wasn't_ an instance of ChaCha20 (or vice versa).

Your response includes an error but there's no stack trace.

> We do not have to check `$this->decrypt instanceof Rijndael` because we are not using any `Rijndael` specific methods.

What if `$this->decrypt` were an instance of `\DateTime`? I mean, I'm not actually suggesting we do a `$this->decrypt instanceof Rijndael` check but I'm just curious how this is different.

If there's a discrepancy between `$this->decrypt` and `$this->decryptName` then it seems likely that we won't be able to decrypt anything from the server anyway. Like if those two variables can get out of sync then that impacts all ciphers and not just chacha20-poly1305@openssh.com.

Reply:
Here's how it decides how to create `$this->decrypt` and `$this->encrypt`:

https://github.com/phpseclib/phpseclib/blob/ea3bee985ce3c96a189b94749c0e79ecd1134455/phpseclib/Net/SSH2.php#L2111-L2148

Here's where that method is called:

https://github.com/phpseclib/phpseclib/blob/ea3bee985ce3c96a189b94749c0e79ecd1134455/phpseclib/Net/SSH2.php#L1952

And then later we assign `$this->decryptName` thusly:

https://github.com/phpseclib/phpseclib/blob/ea3bee985ce3c96a189b94749c0e79ecd1134455/phpseclib/Net/SSH2.php#L1994

If there's a way (short of reflection) for `$this->decryptName` to be chacha20-poly1305@openssh.com whilst `$this->decrypt` _isn't_ an instance of ChaCha20 I'm not seeing it.

Reply:
The point is that Psalm cannot statically determine if `setCounter()` is a valid method because it only knows that `$this->decrypt` is a `SymmetricKey`.

I know that it is impossible for it not to be `ChaCha20` but this is not statically provable. The code was not written with static analysis in mind.

If you insist, I can just suppress this error for now.

Reply:
It can stay but can you add that as a comment? Something to the effect of "_altho it's not possible for $this->decrypt to be anything other than ChaCha20, this check is performed to facilitate static analysis through pslam_".

I can do it myself but won't be able to do so until this evening, after I get home from work!

Thanks! And I appreciate all your PR's! Keep 'em coming!

Reply:
Sure, I'll add the comment when I have a stopping point at work.